Fix centre and square state and add rotation for rectangle selector (#20839)
matplotlib/matplotlib: doc/users/next_whats_new/selector_improvement.rst

Selectors improvement: rotation, aspect ratio correction and add/remove state
-----------------------------------------------------------------------------

The `~matplotlib.widgets.RectangleSelector` and
`~matplotlib.widgets.EllipseSelector` can now be rotated interactively.

> The rotation is enabled or disable by striking the *r* key

Reviewer: ```suggestion
The rotation is enabled or disabled by pressing the *r* key
```

Reviewer 2: And that's also how the actions are labelled internally... but I don't think "strike" is typically used for keyboards (except perhaps the musical variety). Perhaps "typing" or "clicking", but I'd leave that to a native speaker to decide.

> 0 and 45°. The rotation is enabled or disable by striking the *r* key

Reviewer 3: ```suggestion
0 and 45°. The rotation is enabled or disabled by striking the *r* key
```

Reviewer 3: I think we should give a reason for these somewhat particular limits. Please add a sentence:

> The range limits are currently dictated by the implementation.

or similar.

One a side note: Shouldn't it be possible to cover the range -45°...45°, at least there the corner order does not change. One only has to add a sign, which may be possible to infer from the mouse vs start_point location. But this does not have to be in this PR. It's already very big and almost ready to go in.

Author: Good idea, I change to -45 to 45 because it is a easy change!

Reviewer 2: That would be great as it allowed at least to cover the entire range of orientations with some axis tricks!
Could you also add a note what specially was not working with larger angles, so this might be considered in a follow-up as discussed in https://github.com/matplotlib/matplotlib/pull/20839#discussion_r760650455?
In particular, did the problems occur only with specific backends? As I mentioned, I did not notice any serious problems with up to 90 (and in fact now tested up to ±180˚, which effectively allows any number of full rotations). The only obvious issue was, on subsequent resizing of the rotated rectangle, the centre seems to start moving on some kind of trochoid, but whole weird-looking behaviour on resize can of course be avoided by clicking to `'center'` state.

Author: > Could you also add a note what specially was not working with larger angles, so this might be considered in a follow-up as discussed in #20839 (comment)?

I don't remember but I would expect the value of `self._active_handle` to be incorrect when rotated more than -+45, e.g. `'W` is picked while it should be a `S`, etc. 

> The only obvious issue was, on subsequent resizing of the rotated rectangle, the centre seems to start moving on some kind of trochoid, but whole weird-looking behaviour on resize can of course be avoided by clicking to 'center' state.

I have noticed it recently and I think this is related to `_get_aspect_ratio` and it would be good to improve it in another PR.

Reviewer 2: > I don't remember but I would expect the value of self._active_handle to be incorrect when rotated more than -+45, e.g. 'W is picked while it should be a S, etc.

Thanks; this will be a very useful starting point! Yes, I would expect that the point originally labelled `SW` could end up somewhere on the `E` side etc., but could not figure out where that would cause trouble. Any other hints what to watch out for with a follow-up are definitely appreciated.

The motion I think is inevitably somewhat confusing as the fix point for resizing is not rotated along with the bounding box, but that seems quite tricky to address. It is certainly recommendable to use `'center'` mode for resizing of the rotated shapes.

> (default value of 'rotate' in *state_modifier_keys*) or by calling

Reviewer: ```suggestion
('r' is the default key mapped to 'rotate' in *state_modifier_keys*) or by calling
```

> *selector.add_state('rotate')*.

Reviewer: ```suggestion
``selector.add_state('rotate')``.
```
Double ticks are usually used for code snippets

The aspect ratio of the axes can now be taken into account when using the

> "square" state. This is enable by specifying *use_data_coordinates='True'* when

Reviewer: ```suggestion
"square" state. This is enable by specifying ``use_data_coordinates='True'`` when
```

> "square" state. This is enable by specifying *use_data_coordinates='True'* when

Reviewer 2: ```suggestion
"square" state. This is enabled by specifying ``use_data_coordinates='True'`` when
```
As per my general comments, I think this could also do with some more explanation what is meant by "the axes ratio of the axes can be taken into account", i.e. something like "which uses the extents in data coordinates for defining the square aspect ratio".

the selector is initialized.

In addition to changing selector state interactively using the modifier keys
defined in *state_modifier_keys*, the selector state can now be changed

> programmatically using the *add_state* and *remove_state* method.

Reviewer: ```suggestion
programmatically using the *add_state* and *remove_state* methods.
```

.. code-block:: python

    import matplotlib.pyplot as plt
    from matplotlib.widgets import RectangleSelector
    import numpy as np

    values = np.arange(0, 100)

    fig = plt.figure()
    ax = fig.add_subplot()
    ax.plot(values, values)

    selector = RectangleSelector(ax, print, interactive=True,
                                 drag_from_anywhere=True,
                                 use_data_coordinates=True)
    selector.add_state('rotate') # alternatively press 'r' key
    # rotate the selector interactively

    selector.remove_state('rotate') # alternatively press 'r' key

>     selector.add_state('square')

Reviewer 2: I am wondering of some of that documentation and the example could be moved or copied into the docstring so it would continue to be readily available – this look like the most important part of the updated docs, and I am worried  it would get lost in the changelog over time.

> Selectors improvement

Reviewer: Maybe "Rotating selectors" instead?

---------------------

The `~matplotlib.widgets.RectangleSelector` and
`~matplotlib.widgets.EllipseSelector` can now be rotated interactively. The rotation is

> enabled and disabled by pressing the 'r' key.

Reviewer 3: That's a bit ambiguous. Is it one key press (plus release) or do you have to press and hold?

The "square" shape can be defined in data or display coordinate system as defined

> by the *data_coordinates* modifier.

Reviewer 3: From a quick look at the code rotation and data_coordinate both seem to be modifiers. You should then use the same type of description for both - i.e. rather say "d key" instead of *data_coordinates* modifier. Or mention the modifier in both.
